test/functional/forall/atomic-view/tests/test-forall-atomic-view.hpp

```diff
@@ -0,0 +1,145 @@
+#ifndef __TEST_FORALL_ATOMIC_VIEW_HPP__
+#define __TEST_FORALL_ATOMIC_VIEW_HPP__
+
+#include <RAJA/RAJA.hpp>
+#include "RAJA_gtest.hpp"
+
+TYPED_TEST_SUITE_P(SeqForallAtomicViewFunctionalTest);
+
+template <typename T>
+class SeqForallAtomicViewFunctionalTest : public ::testing::Test
+{
+};
+
+#if defined(RAJA_ENABLE_CUDA)
+TYPED_TEST_SUITE_P(CudaForallAtomicViewFunctionalTest);
+
+template <typename T>
+class CudaForallAtomicViewFunctionalTest : public ::testing::Test
+{
+};
+#endif
+
+#if defined(RAJA_ENABLE_HIP)
+TYPED_TEST_SUITE_P(HipForallAtomicViewFunctionalTest);
+
+template <typename T>
+class HipForallAtomicViewFunctionalTest : public ::testing::Test
+{
+};
+#endif
+
+template <typename ExecPolicy,
+          typename AtomicPolicy,
+          typename WORKINGRES,
+          typename T>
+void testAtomicViewBasic( RAJA::Index_type N )
+{
+  RAJA::RangeSegment seg(0, N);
+  RAJA::RangeSegment seg_half(0, N / 2);
+
+  camp::resources::Resource src_res{WORKINGRES()};
+  camp::resources::Resource dest_res{WORKINGRES()};
+
+  T * source = src_res.allocate<T>(N);
+  T * dest = dest_res.allocate<T>(N/2);
+
+#if defined(RAJA_ENABLE_CUDA)
+  cudaErrchk(cudaDeviceSynchronize());
+#endif
+
+#if defined(RAJA_ENABLE_HIP)
+  hipErrchk(hipDeviceSynchronize());
+#endif
+
+  RAJA::forall<RAJA::seq_exec>(seg,
+                               [=](RAJA::Index_type i) { source[i] = (T)1; });
+
+  // use atomic add to reduce the array
+  RAJA::View<T, RAJA::Layout<1>> vec_view(source, N);
+
+  RAJA::View<T, RAJA::Layout<1>> sum_view(dest, N);
+  auto sum_atomic_view = RAJA::make_atomic_view<AtomicPolicy>(sum_view);
+
+
+  // Zero out dest using atomic view
+  RAJA::forall<ExecPolicy>(seg_half, [=] RAJA_HOST_DEVICE(RAJA::Index_type i) {
+    sum_atomic_view(i) = (T)0;
+  });
+
+  // Assign values to dest using atomic view
+  RAJA::forall<ExecPolicy>(seg, [=] RAJA_HOST_DEVICE(RAJA::Index_type i) {
+    sum_atomic_view(i / 2) += vec_view(i);
+  });
+
+#if defined(RAJA_ENABLE_CUDA)
+  cudaErrchk(cudaDeviceSynchronize());
+#endif
+
+#if defined(RAJA_ENABLE_HIP)
+  hipErrchk(hipDeviceSynchronize());
+#endif
+
+  for (RAJA::Index_type i = 0; i < N / 2; ++i) {
+    EXPECT_EQ((T)2, dest[i]);
+  }
+
+  src_res.deallocate( source );
+  dest_res.deallocate( dest );
+}
+
+TYPED_TEST_P(SeqForallAtomicViewFunctionalTest, seq_ForallAtomicViewFunctionalTest)
+{
+  using AExec   = typename camp::at<TypeParam, camp::num<0>>::type;
+  using APol    = typename camp::at<TypeParam, camp::num<1>>::type;
+  using ResType = typename camp::at<TypeParam, camp::num<2>>::type;
+  using DType   = typename camp::at<TypeParam, camp::num<3>>::type;
+  testAtomicViewBasic<AExec, APol, ResType, DType>( 100000 );
+}
+
+REGISTER_TYPED_TEST_SUITE_P( SeqForallAtomicViewFunctionalTest,
+                             seq_ForallAtomicViewFunctionalTest
+                           );
+
+#if defined(RAJA_ENABLE_CUDA)
+GPU_TYPED_TEST_P(CudaForallAtomicViewFunctionalTest, cuda_ForallAtomicViewFunctionalTest)
```

Comment:
Did we ever find out if we still needed to use GPU_TYPED_TEST_P? we aren't having to use it in other functional tests

Reply:
I don't think we do. I would prefer to not use it if we can and eventually remove it.

Reply:
I've reverted this to TYPED_TEST for now. Had a conversation with @davidbeckingsale about this, but it's still not clear if we need GPU_TYPED_TEST. We originally thought that it was written because CUDA had trouble with private test bodies. Upon further inspection of GoogleTest itself, their TestBody() is still private. We can investigate further whether GPU_TYPED_TEST is still required in another PR.

Reply:
It was required because TYPED_TEST turned the block after it into the body of an un-namable method (a constructor, or something I can't remember precisely) such that cuda annotated lambdas inside that block caused NVCC to choke.  If you don't have any cuda device or host/device lambdas in the block, it's not necessary.

Reply:
Ok, all of our current forall tests use host_device decorated lambdas, and they are working with TYPED_TEST. I will try removing GPU_TYPED_TEST from other (older) tests in a separate PR, and if that works I'll remove the macro altogether.

Reply:
These wouldn't run into the problem, because they don't put the lambda inside the body directly in the body of the test.  It's only a problem if you put the lambda syntactically inside the block attached to the TYPED_TEST macro.

Reply:
Ah, that's correct. It fails when there's a host-device lambda in the body of the TYPED_TEST macro. I suppose we can refactor the old tests to avoid this, and still keep the GPU_TYPED macro around just in case.

Reply:
@rchen20 all the tests in the new organization are working without GPU_TYPED_TEST. As we convert more of the old tests and cull out the ones we don't need anymore, we should follow the new pattern and not use the GPU macro. Let's proceed with that plan and revisit this if we run into any issues.